sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -761,72 +795,111 @@ public void processElementForElementAndRestriction(
           continue;
         }
 
-        // Make sure to get the output watermark before we split to ensure that the lower bound
-        // applies to both the primary and residual.
-        KV<Instant, WatermarkEstimatorStateT> watermarkAndState =
-            currentWatermarkEstimator.getWatermarkAndState();
-        SplitResult<RestrictionT> result = currentTracker.trySplit(0);
+        // Attempt to checkpoint the current restriction.
+        HandlesSplits.SplitResult splitResult =
```

Comment:
Why do we want to checkpoint after processElement finishing?

Reply:
The DoFn returned `ProcessContinuation#resume` and not `ProcessContinuation#stop`.

The current restriction within the restriction tracker represents the entire restriction so splitting it provides the remainder and should update the restriction tracker so that the primary is in a done state so that `checkDone` down below can succeed.

Reply:
Got it. Thanks!